### PR TITLE
Add Lookup Table upload support to PAT

### PR DESCRIPTION
**Pull request by matthewcummings**
### Background

It will be helpful for PAT users to be able to perform basic CRUD operations with Lookup Tables.

### Changes

* Added a new lookup_table.py file for the core LUT functionality
* A CLI entrypoint in main.py
* Tests and samples

### Testing

* Unit tests

### CODE REVIEW

File: panther_analysis_tool/lookup_table.py
Comment (rleighton) [inline panther_analysis_tool/lookup_table.py:31]: I would add a comment as to why you need this (to get input bucket).

File: panther_analysis_tool/lookup_table.py
Comment (rleighton) [inline panther_analysis_tool/lookup_table.py:76]: You could reduce this code by adding an endpoint to the LUTs API called `createOrUpdate` that implements this logic. The Go code would be shorter also.

File: panther_analysis_tool/lookup_table.py
Comment (rleighton) [inline panther_analysis_tool/lookup_table.py:177]: I suggest you save the value in a class member so the lambda only gets called once.

File: tests/unit/panther_analysis_tool/test_lookup_tables.py
Comment (rleighton) [inline tests/unit/panther_analysis_tool/test_lookup_tables.py:1]: Do you want this empty file?

Comment (matthewcummings) [reply tests/unit/panther_analysis_tool/test_lookup_tables.py:1]: It's a placeholder, I still need to add tests specifically for lookup tables. Need to get the existing custom schema tests working again first.

File: panther_analysis_tool/lookup_table.py
Comment (matthewcummings) [reply panther_analysis_tool/lookup_table.py:177]: Yeah, was going to set it via @property but it can raise an exception, which feels slightly odd, but maybe it's still Pythonic.

File: panther_analysis_tool/lookup_table.py
Comment (georgepsarakis) [inline panther_analysis_tool/lookup_table.py:179]: I'm not sure if there is need for granting additional permissions to get the function configuration from another Lambda. Alternatively you can create an endpoint in the lookup-tables-api Lambda that just returns the information that `local_file_matches` needs. In general, there is benefit in keeping the client-side simpler (you don't depend so much on version updates for example), so letting the backend handle more complexity can give you more flexibility in the future.

File: panther_analysis_tool/lookup_table.py
Comment (matthewcummings) [reply panther_analysis_tool/lookup_table.py:76]: OK, will move ahead on that front.

File: panther_analysis_tool/lookup_table.py
Comment (matthewcummings) [reply panther_analysis_tool/lookup_table.py:181]: @georgepsarakis yes, good point. However, I could also add the bucket name as an environment variable to the lookup table lambda which is what I was going to do at first but I felt like that would be slightly misleading as the Lambda itself does not need permissions to this bucket. In fact, one of the goals of this PR is to provide an alternate mechanism for uploading files to a lookup table so that we can work around Lambda's payload limit.

But 'saying this out loud', I feel like it's not so bad and is lighter weight to just give the lookup lambda the input bucket name in its configuration (env variables).

Comment (rleighton) [reply panther_analysis_tool/lookup_table.py:181]: Actually, just call this endpoint (https://github.com/panther-labs/panther-enterprise/blob/master/api/lambda/logtypes/models_gen.go#L30)  on the logTypes api (PAT should already have permission) that returns an "upload" s3 path. @shaun-sweet is enhancing the LUTs api to be able to handle these.

Comment (matthewcummings) [reply panther_analysis_tool/lookup_table.py:181]: OK, to clarify you're recommending that I use this new endpoint to pull the name of which bucket? The one used by the log processor or a different one? Also, this would require that I wait until @shaun-sweet has merged his changes to master.

Comment (matthewcummings) [reply panther_analysis_tool/lookup_table.py:181]: @rleighton

File: panther_analysis_tool/lookup_table.py
Comment (matthewcummings) [reply panther_analysis_tool/lookup_table.py]: done

File: panther_analysis_tool/lookup_table.py
Comment (matthewcummings) [reply panther_analysis_tool/lookup_table.py:105]: OK, this code will be going away once I package-ize the s3 signing code.

File: panther_analysis_tool/lookup_table.py
Comment (matthewcummings) [reply panther_analysis_tool/lookup_table.py:103]: going away

File: panther_analysis_tool/lookup_table.py
Comment (darwayne) [inline panther_analysis_tool/lookup_table.py:53]: `[minor]` what do you think about naming this upsert instead?

**Review by darwayne — DISMISSED** [review]
LGTM

File: panther_analysis_tool/lookup_table.py
Comment (matthewcummings) [reply panther_analysis_tool/lookup_table.py:53]: @darwayne There's another file which has an `update_or_create` method so this felt in line with that style. Doesn't matter too much to me either way. Doesn't really feel like an insert/update/upsert since we're not talking to database directly.

But your team owns this repo so tell me what you want. :)

File: panther_analysis_tool/main.py
Comment (lindsey-w) [inline panther_analysis_tool/main.py:296]:
```suggestion
    """
```

File: panther_analysis_tool/lookup_table.py
Comment (lindsey-w) [inline panther_analysis_tool/lookup_table.py:66]: [optional] this would be easier to read if you move the `json.dumps()` into the invoke statement and made this a dictionary with a single field per line

File: tests/unit/panther_analysis_tool/test_lookup_tables.py
Comment (lindsey-w) [inline tests/unit/panther_analysis_tool/test_lookup_tables.py:30]:
```suggestion
    def test_load_invalid_specs_from_folder(self):
```

**Review by lindsey-w — COMMENTED** [review]
Looks good! Might be nice to include additional unit tests for the LookupTable methods?

File: panther_analysis_tool/lookup_table.py
Comment (matthewcummings) [reply panther_analysis_tool/lookup_table.py:66]: I did that so that I could add the logging.debug() line. Can remove if you don't think that helps. @lindsey-w

Comment (matthewcummings) [reply panther_analysis_tool/lookup_table.py:63]: @lindsey-w I considered adding more tests but as unit tests, we can't really test much with this code other than that it parses responses output correctly. Also, I have integration tests in the backend which cover the API functionality.

It would be nice to have a framework in place that ran PAT tests against an actual backend but it would take some effort to build that out.

**Review by lindsey-w — DISMISSED** [review]

File: panther_analysis_tool/main.py
Comment (georgepsarakis) [inline panther_analysis_tool/main.py:430]: Maybe use debug level?

File: panther_analysis_tool/lookup_table.py
Comment (georgepsarakis) [inline panther_analysis_tool/lookup_table.py:166]: Same note about logging presigned URLs as above, in case you find it important.

**Review by georgepsarakis — COMMENTED** [review]
Nice work!

File: panther_analysis_tool/lookup_table.py
Comment (matthewcummings) [reply panther_analysis_tool/lookup_table.py:1]: I could but my team doesn't "own" this code + it's public + even if we did own this code, I'm really not trying to overhaul everything because my first thought is the several other files in this directory should also be their own packages.

File: panther_analysis_tool/main.py
Comment (matthewcummings) [reply panther_analysis_tool/main.py:430]: I would say no, I usually lean towards verbosity with tools like PAT which run in CI/CD pipelines since it makes debugging easier and this is not super verbose in my view.

File: panther_analysis_tool/lookup_table.py
Comment (matthewcummings) [reply panther_analysis_tool/lookup_table.py:87]: This is copy/pasta from other parts of the codebase. In many cases I've just been keeping with the "style" of this code.

File: panther_analysis_tool/lookup_table.py
Comment (matthewcummings) [reply panther_analysis_tool/lookup_table.py:1]: OK, I'll do it. :)

File: panther_analysis_tool/lookup_table.py
Comment (matthewcummings) [reply panther_analysis_tool/lookup_table.py:82]: Lookup tables names are unique.

Comment (georgepsarakis) [reply panther_analysis_tool/lookup_table.py:82]: Aha thanks for clarifying!

File: panther_analysis_tool/main.py
Comment (georgepsarakis) [reply panther_analysis_tool/main.py:430]: Sure, makes sense.

File: panther_analysis_tool/lookup_table.py
Comment (darwayne) [reply panther_analysis_tool/lookup_table.py]: @matthewcummings got it. Anytime I see a function that is only responsible for creating or updating I recommend naming it upsert since it is concise an encapsulates both create and update. create is synonymous to insert. So upsert doesn't necessarily imply what your underlying storage mechanism is.

**Review by darwayne — DISMISSED** [review]

File: panther_analysis_tool/main.py
Comment (georgepsarakis) [inline panther_analysis_tool/main.py:67]: Reminder to include the package in `setup.py` (see [here](https://github.com/panther-labs/panther_analysis_tool/blob/199c2d294e25b9ef332045996544974ab970ed87/setup.py#L24)) so that it is included in the PyPI release.

**Review by darwayne — APPROVED** [review]

**Review by darwayne — APPROVED** [review]

**Pull request merged**
